source/common/upstream/load_balancer_impl.cc

```diff
@@ -267,6 +303,23 @@ bool ZoneAwareLoadBalancerBase::earlyExitNonLocalityRouting() {
   return false;
 }
 
+HostConstSharedPtr LoadBalancerBase::chooseHost(LoadBalancerContext* context) {
+  HostConstSharedPtr host;
+  size_t max_attempts = context ? context->hostSelectionRetryCount() + 1 : 1;
+  for (size_t i = 0; i < max_attempts; ++i) {
+    host = chooseHostOnce(context);
+
+    // If host selection failed or the host is accepted by the filter, return.
+    // Otherwise, try again.
+    if (!host || !context || context->postHostSelectionFilter(*host)) {
```

Comment:
Should we allow a filter to decide whether to try again if host is nullptr? Not sure, just throwing that out there.